### ISSUE TITLE
perf: close fresh-tape ~233ms gap to PyTorch (per-op pool + plan cache + compiled-IL backward)

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#338]
## Summary

After PR #331 (Phase 3 wins) and the OpenBLAS-default fix in `perf/327-parallel-transposed-gemm`, the **persistent-tape** Transformer training path hits the PyTorch CPU stretch target of 50 ms/step. The **fresh-tape** consumer-facing path remains at **~291 ms/step** — a **5–6× gap to the 50 ms target** that is NOT GEMM compute (OpenBLAS is now active for both paths). The remaining gap is autodiff dispatch overhead specific to the fresh-tape lifecycle. This issue tracks closing it.

Cross-references:
- #327 (parent perf issue) — closed by PR #331 + OpenBLAS fix for the persistent-tape path.
- PR #331 — landed `MatMulBackward` collapse-to-2D, optimizer parallelization, allocation reductions; documented this fresh-tape gap as "remaining work".

## Measurement

Hardware: AMD Ryzen 9 3950X (16C/32T), Windows 11, .NET 10, AiDotNet.Tensors at the tip of the `perf/327-parallel-transposed-gemm` branch (PR #331 + OpenBLAS).

Test: `Issue327TransformerTrainPerfTests.Issue327_Transformer_FreshTape_TrainStep_DoesNotExplodeCatastrophically`. Config d=128, L=4, B=32, ctx=64, V=256 (test scope).

3-run median wall-time per training iteration:

| Path | Pre-OpenBLAS | Post-OpenBLAS | PyTorch CPU MKL target |
|---|---|---|---|
| Persistent-tape (compiled-chain cache active) | 103 ms | **58 ms** | ~50 ms |
| **Fresh-tape (no cache)** | **354 ms** | **291 ms** | **~50 ms** |

The 233 ms remaining gap on fresh-tape is dominated by per-call autodiff lifecycle overhead, NOT GEMM compute. OpenBLAS is dispatched identically on both paths; the persistent-tape path's `RebindablePlanCache` skips the per-call backward graph walk and intermediate allocation, exposing the fresh-tape path's structural overhead as the residual.

## What's already done (PR #331 + this branch)

- ✅ `MatMulBackward` collapse-to-2D (parallel 2D GEMM via `engine.TensorMatMul`) — 287 → 162 ms
- ✅ `ReduceSumBackward` skip zero+fill (`RentZeroed → RentUninitialized`) — ~5–10 ms typical
- ✅ Tall-thin K threshold 256 → 512 — captures FFN-down backward — 150 → 140 ms
- ✅ `GradientsScope<T>` (explicit grad-pool lifecycle) — −45% bwd allocation, −50% GC
- ✅ Parallel optimizer step — 7 → 3.3 ms
- ✅ `RebindablePlanCache<T>` (cross-tape backward-plan cache) — engages on persistent-tape, **NOT engaged on fresh-tape**
- ✅ OpenBLAS native dispatch (this branch) — closes the GEMM compute gap entirely

### 1. Per-op backward intermediate pooling with tape-pinning (~10-20 ms expected)

Backward functions currently allocate per-call (e.g., `bT2`/`aT2` in MatMulBackward's transpose path, GELU saved-input copies, ReLU mask scratch). Pool-returning these is unsafe today because a recorded backward op may still hold a reference. The fix is **tape-pinning**: add `bool _pinnedByTape` on `Tensor<T>`, set when a tensor is recorded as input to a tape entry, checked in `TensorPool.Return`.

Scope: ~50 backward functions in `BackwardFunctions.cs`. Each gets a pool-rent at the top + a `Return` in `finally`, gated on the pin flag.

References:
- PR #331 attempted "Pool-return for MatMul backward `bT2`/`aT2`" → broke `Hvp_MatchesHessianTimesVec` (SumToScalarTensor aliasing). Pin-flag fixes this.
- PR #331 `GradientsScope<T>` is the prototype; this generalizes the pattern to per-op scratch.

Estimated savings: 10-20 ms/iter on the fresh-tape path. Per-op allocation costs are typically 0.1–0.5 ms × ~50 op-calls/iter.

### 2. Cache fresh-tape backward plans by op-pattern hash (~50-100 ms expected)

The fresh-tape path does a full backward graph walk (DFS topo-sort, step array construction, dispatch loop) every call. The persistent-tape path's `RebindablePlanCache<T>` already caches the **plan** (an int[] of entry indices) by an op-pattern hash. Extend the cache so a fresh tape with an identical op pattern (same forward function called with same shapes) **also** hits the cache after a one-time DFS.

Key design point: the cache must be **fresh-tape-safe** — it can hold no Tensor references (those are per-tape), only entry-index permutations. PR #331's `RebindablePlanCache` is already structured this way; the gate is just lifted from "persistent only" to "any non-create-graph tape".

Risks:
- Cache key collision on differing forwards with same op pattern: hash includes shapes + op-type sequence, so collisions are unlikely. Add a defensive shape-equality check on cache hit before dispatch.
- Memory: cache entries are small (int[]), and key is a HashCode.Combine — bounded by # of distinct forward patterns.

Estimated savings: 50-100 ms/iter. The fresh-tape graph walk is currently ~17 MatMul + ~244 GELU + ~244 Slice + ~61 ReduceSum + ~hundreds-of-other backward dispatches × ~30 µs overhead/dispatch = the bulk of the 233 ms.

### 3. Compiled-IL backward (~30-50 ms expected, long-term)

Equivalent to torch.compile / Inductor: at first fresh-tape backward call, emit IL that hard-codes the dispatch sequence as a single method (eliminating the per-step delegate invocation overhead). Compiled methods get JITted to tight code with no dispatch tax. Cached by the same op-pattern hash as item 2.

Scope: write a `BackwardPlanCompiler` that takes a `RebindablePlan` + tape state, emits `DynamicMethod` IL calling each backward function inline. Roughly 1500 LoC. Validated against the existing dispatcher with seed-by-seed gradient parity.

Estimated savings: 30-50 ms/iter on top of items 1+2.

## Projected wall-time after closing this issue

| Stage | Fresh-tape wall | Cumulative reduction |
|---|---|---|
| Today (post-OpenBLAS) | 291 ms | — |
| + Per-op pooling (item 1) | ~275 ms | −16 ms (5%) |
| + Plan cache for fresh-tape (item 2) | ~200 ms | −75 ms (32% from today) |
| + Compiled-IL backward (item 3) | **~50 ms** | **−241 ms (83% from today)** |

After all three items, fresh-tape sits at the persistent-tape level (~50 ms), and PyTorch CPU MKL parity is achieved on the consumer path.

## Suggested ordering

1. Item 1 (per-op pooling with tape-pinning) — ~1-2 weeks, lowest risk, lays groundwork. Decoupled from item 2.
2. Item 2 (fresh-tape plan cache) — ~3-5 days after item 1 lands, builds on `RebindablePlanCache`'s existing infrastructure.
3. Item 3 (compiled-IL backward) — ~1-3 months, builds on item 2's cache layer. Highest reward, also highest complexity.

Items 1 + 2 are sufficient to bring fresh-tape into the ≤ 200 ms range (PR #331's stated close criterion at the looser bound). Item 3 is what hits the ≤ 50 ms PyTorch stretch target.

## Close criteria

- Fresh-tape `Issue327_Transformer_FreshTape_TrainStep_DoesNotExplodeCatastrophically` wall-time ≤ 100 ms on a 16-core x64 host.
- Stretch (PyTorch parity): ≤ 50 ms.
- All 802+ tape/autodiff/MatMul/Backward tests still pass.
- Sources w/ gradients still 17/17.
- No regressions in `Issue327_Transformer_PersistentTape_TrainStep_BelowIssueCloseTarget` (currently at 58 ms median).

## Invocation

```bash
# Fresh-tape measurement (this issue's target):
dotnet test tests/AiDotNet.Tensors.Tests --filter "FullyQualifiedName~Issue327_Transformer_FreshTape_TrainStep_DoesNotExplodeCatastrophically"

# Backward timing breakdown:
AIDOTNET_BWD_TIMING=1 dotnet run -c Release --project tests/AiDotNet.Tensors.Benchmarks -- --327-transformer
```
